File: chat-client/src/types/websocket.ts
import { Message } from "./message"

export type WebSocketAction = 'send_message' | 'send_message_error' | 'new_message' | 'broadcast_error'

export interface WebSocketMessage<T = unknown> {
  action: WebSocketAction
  message_id?: string
  data: T
}

export interface SendMessageData {
  content: string
  chat_id: number
}

export interface NewMessageData {
  message: Message
}

export interface SendMessageErrorData {
  error: string
  message_id: string
} 

File: chat-server/models/websocket.go
package models

import (
	"encoding/json"
	"log"
)

type WebSocketMessage struct {
	Action    string `json:"action"`
	// random uuid for the message
	MessageID string `json:"message_id"`
	Data      any    `json:"data"`
}

func (message *WebSocketMessage) Encode() []byte {
	encoding, err := json.Marshal(message)
	if err != nil {
		log.Println(err)
	}

	return encoding
}

type SendMessageData struct {
	ChatID  uint   `json:"chat_id"`
	Content string `json:"content"`
}